OpenTable: Accept more embed options

scruffian opened the pull request:
#### Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->
* This broadens the scope of what kinds of embed the OpenTable block will accept
* It will now take an embed code, a URL, or a list of selected restaurants.

#### Testing instructions:
* Add an OpenTable block
* Check that you can still search for restaurants and embed a widget based on the results
* Check that you can copy/paste the embed code from this page into the embed code form: 
https://www.opentable.com/widget/reservation/preview?rid=1&lang=en-US
* Check that you can copy/paste the URL from this page into the embed code form: https://www.opentable.com/widget/reservation/preview?rid=1&lang=en-US
* Also test other forms of URL, for example:
- https://www.opentable.com/restref/client/?restref=412810&lang=en-US&ot_source=Restaurant%20website&corrid=e413926b-0352-46d6-a8d8-d1d525932310
- https://www.opentable.com/restref/client/?restref=1&lang=es-MX&ot_source=Restaurant%20website&corrid=09f44cc6-f0cb-4e98-9298-f4ba8cc20183 (uses a different language)
- https://www.opentable.com/restref/client/?rid=1&corrid=010a3136-569e-42a5-a381-e111887b4cf5
* Also test that these URLs work when pasted directly into the editor
* Make sure that other types of embed still work.


#### Proposed changelog entry for your changes:
<!-- Please do not leave this empty. If no changelog entry needed, state as such. -->
* Accept URLs when embedding an OpenTable block

scruffian committed:
get the restaurant id from the URL if it's available

scruffian committed:
Accept URLs pasted into the editor

scruffian commented on extensions/blocks/opentable/utils.js:
```diff
@@ -0,0 +1,62 @@
+
+	let domain = searchParams.get( 'domain' );
+	if ( ! domain || domain.length === 0 ) {
+		domain = 'com';
```
We might be able to get this from the URL, but I don't think it matters.

jeherve reviewed (changes requested):
That cleans things up quite nicely. 

I think we'd benefit from a test for `getAttributesFromEmbedCode` here, to avoid any regressions in the future, just like we've done for Calendly:
https://github.com/Automattic/jetpack/blob/2c3dbd93a0cc735df5593a147532cf5de02627ba/extensions/blocks/calendly/test/utils.js

scruffian commented:
I added unit tests for this function. In the process I realised that it wasn't actually working correctly - we weren't picking up the correct true/false values from the widget code. I have fixed this now, but to do so I had to change the `iframe` and `newtab` attributes from `Boolean`s into `string`s. 

The only issue I see with this is that users who have previously changed these settings to something other than the default will lose them if they edit a post with a block in. I'm not sure its worth adding code to account for this. What do you think?

jeherve commented:
> I'm not sure its worth adding code to account for this. What do you think?

In the past we have. You can check the VideoPress for an example:
https://github.com/Automattic/Jetpack/commit/57ed23f5e9e672d7c59324ab529b5284d619a253

That's a lot for a small change, though. Could we make the type change before we get to the attributes instead?

scruffian commented:
Based on some useful feedback from @pablinos I did this a different way; we are keeping the boolean attributes and just casting them in the validation helper.

scruffian committed:
Adds tests for getValidatedAttributes

scruffian commented:
I also added some tests for `getValidatedAttributes`

pablinos reviewed (commented):
Looks good! Just a couple of comments about the regex and error handling. Testing this threw up another bug that I've tried to address in #14678

scruffian commented on extensions/blocks/opentable/utils.js:
```diff
@@ -0,0 +1,80 @@
+		restaurantId = searchParams.getAll( 'restref' );
+	}
+	if ( ! restaurantId || restaurantId.length === 0 ) {
+		return;
```
Yeah we should probably, but I think that would be good for a different PR :)

scruffian commented on extensions/blocks/opentable/utils.js:
```diff
@@ -0,0 +1,80 @@
+export const embedRegex = /< *script[^>]*src *= *["']?([^"']*)/i;
```
Seems unlikely, but this is the REGEX we already had in there...

pablinos replied:
Oh yes, so it is! Seems odd, but if it's working :)

jeherve reviewed (changes requested):
This is nice. I wish we could support regular URLs as well, such as https://www.opentable.com/r/bistro-deryne-budapest?corrid=dae839dd-c3f2-42c1-8796-91668437dbf7 , but that's for another time I guess :)

Before we merge I only have one nitpick to save us some confusion later on.

pablinos committed:
Change the regex to use \s for all whitespace

pablinos committed:
Add tests for invalid embed codes and URLs

This also adds a check that the host of the URL includes OpenTable

pablinos committed:
Converted spaces to tabs

pablinos commented on extensions/shared/test/get-validated-attributes.js:
```diff
@@ -0,0 +1,151 @@
+const urlValidator = url => ! url || url.startsWith( 'http' );
+
+const defaultAttributes = {
+    boolean: {
```
Nice spot! I've updated that.

pablinos commented:
> I wish we could support regular URLs as well, such as https://www.opentable.com/r/bistro-deryne-budapest?corrid=dae839dd-c3f2-42c1-8796-91668437dbf7

Yeah, I don't think we've got a way of getting the restaurant ID from those URLs currently. although it is available in the HTML source of that page, so we could maybe do a request on the server and grab this vie an API endpoint.

jeherve reviewed (approved):
This should be good to go now.

scruffian commented:
r203007-wpcom
